### PR TITLE
Introduce CliAction, make it possible for every Symbol to define one

### PR DESCRIPTION
**Pull request by adamsitnik**
Overall it was way easier than I expected.

This PR:

1. Replaces `ICommandHandler` with `CliAction`
2. Renames `Command.SetHandler` with `Command.SetAction`. 
3. Changes the exit code approach: every successful program should return 0, so to avoid verbosity command actions provided by `Command.SetAction` do not need to return an int. In case somebody wants to reuturn an int, they need to implement  their own `CliAction` type (pay for play) 
5. Extends Options and Directives with `Action` property, but not `SetAction` methods (no need to, as it should be rare to have active options and directives that don't need custom types)
6. It does not expose the new action types, I want to do it in a separate PR, where I am going to write tests for them and move the configuration to these types.

Contributes to #2071

### CODE REVIEW

File: src/System.CommandLine.Tests/OptionTests.cs [hunk 1]
```diff
@@ -346,5 +346,38 @@ public void Option_of_enum_can_limit_enum_members_as_valid_values()
                 .Should()
                 .BeEquivalentTo(new[] { $"Argument 'Fuschia' not recognized. Must be one of:\n\t'Red'\n\t'Green'" });
         }
+
+        [Fact]
+        public void Every_option_can_provide_a_handler_and_it_takes_precedence_over_command_handler()
```
Comment (jonsequitur) [inline src/System.CommandLine.Tests/OptionTests.cs:351]: What if two different options having handlers are both present?

**Comment by jonsequitur** [comment]
>  I am not sure whether it should be allowed for Arguments, currently the argument handlers are ignored.

I wasn't expecting this to be available on `Argument` and I'm not sure how to reason about it. I don't believe I've seen a CLI where I would model something this way.

File: src/System.CommandLine.Tests/OptionTests.cs (hunk 1 above)
Comment (adamsitnik) [reply src/System.CommandLine.Tests/OptionTests.cs:354]: We have two possibilities:

- stop on the first active option
- let the last one to be the effective (this is what we already do for commands)

@jonsequitur choosing the last one is the most intuitive way for me. What do you think?

**Comment by adamsitnik** [comment]
> I wasn't expecting this to be available on Argument and I'm not sure how to reason about it. I don't believe I've seen a CLI where I would model something this way.

That is very true. I've applied following  changes:

- Command, Option and Directive expose `CliAction Action` property
- only Command exposes `SetAction` methods (for options and directives it's too rare to pollute the public API surface)

File: src/System.CommandLine.Tests/OptionTests.cs (hunk 1 above)
Comment (jonsequitur) [reply src/System.CommandLine.Tests/OptionTests.cs:352]: We could also make it a parse error. Of course certain options (like `-h`) ignore that and the action still runs.

**Comment by tmds** [comment]
@adamsitnik in https://github.com/dotnet/command-line-api/issues/2046 you said: all handlers should return an `int`.

And here the design is:

> Command.SetAction do not need to return an int. In case somebody wants to reuturn an int, they need to implement their own CliAction type (pay for play)

Failures are often expected in command line applications.

Without the capability to return an `int`, command handler set using `SetAction` are forced to `throw` for failure.

This lead to a bad user experience.
And, throwing exceptions for expected failures is a performance anti-pattern.

**Pull request merged**